Allow shortening lifetime in CoerceUnsized for &mut

theemathas opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/149219)*
<!-- homu-ignore:end -->

This modifies the &mut -> &mut CoerceUnsized impl so that, it can shorten the lifetime.

Note that there are already two impls that allow shortening the lifetime like this (the &mut T -> &U and the &T -> &U impls). So this change makes the impls consistent with each other.

I initially tried to also do the same to the CoerceUnsized impl for core::cell::{Ref, RefMut}. However, this can't be done because Ref and RefMut "store" the lifetime and the data in different fields, and CoerceUnsized can only coerce one field.

This change has an effect on stable code, since it allows shortening lifetimes inside invariant types via a &mut -> &mut unsize coercion.

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-20-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/19598190078/job/56125818510) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/56125818510) [(plain)](https://github.com/rust-lang/rust/commit/6bfef635d4b78888d73bc65c076f6c3613033942/checks/56125818510/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
[RUSTC-TIMING] build_script_build test:false 0.323
error[E0375]: implementing `CoerceUnsized` does not allow multiple fields to be coerced
    --> library/core/src/cell.rs:1792:52
     |
1792 | impl<'a, 'b: 'a, T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<Ref<'a, U>> for Ref<'b, T> {}
     |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^
     |
note: the trait `CoerceUnsized` may only be implemented when a single field is being coerced
    --> library/core/src/cell.rs:1584:5
     |
1584 |     value: NonNull<T>,
     |     ^^^^^^^^^^^^^^^^^
1585 |     borrow: BorrowRef<'b>,
     |     ^^^^^^^^^^^^^^^^^^^^^

error[E0375]: implementing `CoerceUnsized` does not allow multiple fields to be coerced
    --> library/core/src/cell.rs:2098:52
     |
2098 | impl<'a, 'b: 'a, T: ?Sized + Unsize<U>, U: ?Sized> CoerceUnsized<RefMut<'a, U>> for RefMut<'b, T> {}
     |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
note: the trait `CoerceUnsized` may only be implemented when a single field is being coerced
    --> library/core/src/cell.rs:2066:5
     |
2066 |     value: NonNull<T>,
     |     ^^^^^^^^^^^^^^^^^
2067 |     borrow: BorrowRefMut<'b>,
     |     ^^^^^^^^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0375`.
[RUSTC-TIMING] core test:false 23.533
error: could not compile `core` (lib) due to 2 previous errors
```

</details>

theemathas commented:
I don't quite understand the errors here...

zachs18 commented:
IIUC, the errors are because when implementing `CoerceUnsized` for ADTs, the non-coerced fields (i.e. the fields other than the pointer field) must be *exactly* the same type between the source and destination instantiations in the current implementation.

https://github.com/rust-lang/rust/blob/94b49fd998d6723e0a9240a7cff5f9df37b84dd8/compiler/rustc_hir_analysis/src/coherence/builtin.rs#L485-502 (the error is produced on line 557, lines 485-492 describe the check)

Since `BorrowRefMut<'a>` is not *exactly* the same type as `BorrowRefMut<'b>` (as per `Ty::eq`), this isn't allowed. The comment starting on line 552 (just before the equality check) is also probably relevant https://github.com/rust-lang/rust/blob/94b49fd998d6723e0a9240a7cff5f9df37b84dd8/compiler/rustc_hir_analysis/src/coherence/builtin.rs#L522-L533

theemathas committed:
Allow shortening lifetime in CoerceUnsized for &mut

This modifies the &mut -> &mut CoerceUnsized impl so that, it can
shorten the lifetime.

Note that there are already two impls that allow shortening the lifetime
like this (the &mut T -> &U and the &T -> &U impls). So this change
makes the impls consistent with each other.

I initially tried to also do the same to the CoerceUnsized impl for
core::cell::{Ref, RefMut}. However, this can't be done because
Ref and RefMut "store" the lifetime and the data in different fields,
and CoerceUnsized can only coerce one field.

This change has an effect on stable code, since it allows shortening
lifetimes inside invariant types via a &mut -> &mut unsize coercion.

scottmcm commented:
TBH, I feel quite unqualified to review the soundness of this.  Maybe someone from types feels more confident?

theemathas commented:
I already tried and couldn't find a case where there's user-observable behavior change, other than directly writing a trait bound on CoerceUnsized or something similar.

I wanted to make this change in order to make the `& -> &` and `&mut -> &` and `&mut -> &mut` impls consistent with each other, as a matter of principle. Although I don't have a particular opinion on whether the lifetime-shortening should be allowed in all three impls or none of the three impls.

If we do stabilize `CoerceUnsized` at some point, an inconsistency between the three impls might result in weird borrow-check errors that are hard to explain to a user who's trying to be generic over types that implement `CoerceUnsized`.

theemathas commented:
@jackh726 There seems to be no user-observable effect due to this change in our test suite.

I tried removing the lifetime shortening on the `& -> &` and `&mut -> &` impls.

<details>
<summary>Diff of changes</summary>

```diff
diff --git a/library/core/src/ops/unsize.rs b/library/core/src/ops/unsize.rs
index f0781ee01fd..e39dded51d4 100644
--- a/library/core/src/ops/unsize.rs
+++ b/library/core/src/ops/unsize.rs
@@ -42,7 +42,7 @@ pub trait CoerceUnsized<T: PointeeSized> {
 impl<'a, T: PointeeSized + Unsize<U>, U: PointeeSized> CoerceUnsized<&'a mut U> for &'a mut T {}
 // &mut T -> &U
 #[unstable(feature = "coerce_unsized", issue = "18598")]
-impl<'a, 'b: 'a, T: PointeeSized + Unsize<U>, U: PointeeSized> CoerceUnsized<&'a U> for &'b mut T {}
+impl<'a, T: PointeeSized + Unsize<U>, U: PointeeSized> CoerceUnsized<&'a U> for &'a mut T {}
 // &mut T -> *mut U
 #[unstable(feature = "coerce_unsized", issue = "18598")]
 impl<'a, T: PointeeSized + Unsize<U>, U: PointeeSized> CoerceUnsized<*mut U> for &'a mut T {}
@@ -52,7 +52,7 @@ impl<'a, T: PointeeSized + Unsize<U>, U: PointeeSized> CoerceUnsized<*const U> f

 // &T -> &U
 #[unstable(feature = "coerce_unsized", issue = "18598")]
-impl<'a, 'b: 'a, T: PointeeSized + Unsize<U>, U: PointeeSized> CoerceUnsized<&'a U> for &'b T {}
+impl<'a, T: PointeeSized + Unsize<U>, U: PointeeSized> CoerceUnsized<&'a U> for &'a T {}
 // &T -> *const U
 #[unstable(feature = "coerce_unsized", issue = "18598")]
 impl<'a, T: PointeeSized + Unsize<U>, U: PointeeSized> CoerceUnsized<*const U> for &'a T {}
```

</details>

The only test failure was in [tests/ui/error-codes/E0476.rs](https://github.com/rust-lang/rust/blob/be4794c78beca8a88f0a1ee583f3d4b5c321e5d8/tests/ui/error-codes/E0476.rs). And it's only a diagnostics change due to rustc outputting the contents of the impl in question in the diagnostics.

<details>
<summary>Diff of change in test output</summary>

```diff
diff --git a/tests/ui/error-codes/E0476.next.stderr b/tests/ui/error-codes/E0476.next.stderr
index 454dbecc7d0..c8fd40d144c 100644
--- a/tests/ui/error-codes/E0476.next.stderr
+++ b/tests/ui/error-codes/E0476.next.stderr
@@ -5,8 +5,8 @@ LL | impl<'a, 'b, T, S> CoerceUnsized<&'a Wrapper<T>> for &'b Wrapper<S> where S
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `core`:
-           - impl<'a, 'b, T, U> CoerceUnsized<&'a U> for &'b T
-             where 'b: 'a, T: Unsize<U>, T: ?Sized, U: ?Sized;
+           - impl<'a, T, U> CoerceUnsized<&'a U> for &'a T
+             where T: Unsize<U>, T: ?Sized, U: ?Sized;

 error[E0476]: lifetime of the source pointer does not outlive lifetime bound of the object type
   --> $DIR/E0476.rs:11:1
diff --git a/tests/ui/error-codes/E0476.old.stderr b/tests/ui/error-codes/E0476.old.stderr
index 454dbecc7d0..c8fd40d144c 100644
--- a/tests/ui/error-codes/E0476.old.stderr
+++ b/tests/ui/error-codes/E0476.old.stderr
@@ -5,8 +5,8 @@ LL | impl<'a, 'b, T, S> CoerceUnsized<&'a Wrapper<T>> for &'b Wrapper<S> where S
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: conflicting implementation in crate `core`:
-           - impl<'a, 'b, T, U> CoerceUnsized<&'a U> for &'b T
-             where 'b: 'a, T: Unsize<U>, T: ?Sized, U: ?Sized;
+           - impl<'a, T, U> CoerceUnsized<&'a U> for &'a T
+             where T: Unsize<U>, T: ?Sized, U: ?Sized;

 error[E0476]: lifetime of the source pointer does not outlive lifetime bound of the object type
   --> $DIR/E0476.rs:11:1
```

</details>

jackh726 commented:
> I just realized: This change actually _does_ have a visible effect in stable rust.

Please add this as a test here.

jackh726 commented:
Okay, discussion on Zulip hasn't raised any issues here. Going to fcp merge for types here. I'm going to ping @rust-lang/libs-api, but I think this is probably "just okay" to have as a types team PR. This isn't really a new "API surface" as much as a change to what coercions we accept.

---

This impl change allows (as an example) the following to compile:

```rust
// currently compiles
fn with_trait<'a: 'b, 'b>(x: Cell<&'a dyn Trait>) -> Cell<&'b dyn Trait> {
    x
}

// currently errors, will compile with this PR
fn with_trait_mut<'a: 'b, 'b>(x: Cell<&'a mut dyn Trait>) -> Cell<&'b mut dyn Trait> {
    x
}
```

@rfcbot merge types

theemathas commented:
I'm not sure whether this will cause problems with user-defined `CoerceUnsized` impls in the future (once we stabilize that). For example:

```rust
#![feature(coerce_unsized)]

use std::ops::CoerceUnsized;
use std::marker::PhantomData;

pub struct Inv<P>(PhantomData<fn(P) -> P>, P);

impl<T: CoerceUnsized<U>, U> CoerceUnsized<Inv<U>> for Inv<T> {}

pub trait Trait {}

// compiles
pub fn works<'a>(x: Inv<&'static dyn Trait>) -> Inv<&'a dyn Trait> {
    x
}

// currently errors, will compile after this PR
pub fn doesnt_work_yet<'a>(x: Inv<&'static mut dyn Trait>) -> Inv<&'a mut dyn Trait> {
    x
}

// errors
pub fn fails<'a>(x: Inv<&'static i32>) -> Inv<&'a i32> {
    x
}
```

In other words, allowing this shortening of lifetimes may potentially change the safety requirements / reasoning needed for one to implement the `CoerceUnsized` trait in the future. This is because allowing more coercions means that implementors of `CoerceUnsized` can assume fewer things as invariants.

I am unable to figure out a nice way to state what invariants a `CoerceUnsized` impl may break. (In other words, I'm unsure what safety invariants a type could impose on its private fields if the type implements `CoerceUnsized`.)

As a result, I am unable to reason whether this PR is a good idea or a bad idea. (I don't have concrete concerns. I just find it hard if not impossible to give a proper justification on whether this PR is doing the right thing.)

An alternative to this PR is to go the other way: disallow lifetime-shortening coercions on shared references, to make them consistent with exclusive references.

Does the interaction with future implementors of `CoerceUnsized` make this PR require a decision from T-libs-api?

lcnr commented:
I personally feel fine with the issue described in the above comment. I cannot imagine a user writing 

```rust
impl<T: CoerceUnsized<U>, U> CoerceUnsized<MyType<U>> for MyType<T> {}
```
while relying on some "does not shorten lifetimes" of that `CoerceUnsized` impl. Being able to shorten lifetimes this way is generally desirable. One unsound thing would be a type whose drop impl writes its only field to some side-table. That way you can own a value which still has to be treated as *shared*.

We should add a comment to `CoerceUnsized` explaining how there are impls which shorten lifetimes, so if you implement `CoerceUnsized` for a type which has shared ownership of `T`, it would be unsound.

This ties into there being a difference `MutableAccess` and `MutableSharedAccess` wrt how variance should be computed, which I've also alluded to in https://rust-lang.zulipchat.com/#narrow/channel/326866-t-types.2Fnominated/topic/.23149219.3A.20Allow.20shortening.20lifetime.20in.20CoerceUnsized.20for.20.26mut/near/582743558

BoxyUwU commented:
I've not looked very closely at this but I'm happy to just trust y'all here so gonna tick my box

theemathas commented:
I've added a test.

theemathas commented:
@jackh726 This PR is still waiting for approval.

jackh726 commented on tests/ui/coercion/unsize-lifetime-shrinking.rs:
```diff
@@ -0,0 +1,93 @@
+// Even though a type being invariant prevents shortening lifetimes via subtyping,
```
I understand the point of the macros here is to be able to test many different possibilities/combinations. But, this test is terribly confusing. I would *much* rather have a longer test with duplicated functions where it's clear what things error, rather than have them obfuscated by macros.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149219/83adf7e080d0fe4f0970c00eac2976a767dbd042..f4e56611fa53272accc2ca85ed1f86ffb4b8c62c)*

theemathas committed:
Add test for shortening lifetimes via unsize coercions.

theemathas commented:
@rustbot ready

theemathas commented:
@jackh726 This PR is still waiting for approval.

theemathas commented:
@rustbot reroll

camelid commented:
r? @jackh726 

Re-assigning Jack since he has context for this and I do not.

jackh726 commented:
@bors r+ rollup

theemathas commented:
@bors r=jackh726 rollup
